### PR TITLE
Speedup gene overview

### LINKED ISSUE #357: Gene page performance
Right now, the gene page takes a minute or so to load. This becomes unfeasible when working through a list of many partially covered genes.

Testing it out, it took ~3 minutes for one gene to load:

gene_overview_loading_time

Btw I am happy to help out with hands-on contribution, both through code and testing. Just let me know if needed / wanted, and we can coordinate :)

### PR DESCRIPTION
**Pull request by northwestwitch**
### Added/Changed/Fixed
- Closes #357 - refactored and simplified code to avoid many operations

### How to test
- [x] Deploy on stage
- [x] Check gene overview page

### Expected outcome
- Page should load faster than with current main branch

## Review
- [x] Tests executed by CR
- [ ] "Merge and deploy" approved by

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,8 @@
 - Updated tests workflow to cargo install the latest d4tools from git (master branch)
 - Computing coverage completeness stats using d4tools `perc_cov` stat function (much quicker reports)
 - Moved functions computing the coverage stats to a separate `meta/handle_coverage_stats.py` module
+- Refactored gene code computing stats for gene overview
+- Display only transcripts with associated refseq mRNA on gene overview
```
Comment (northwestwitch) [inline CHANGELOG.md:18]: Or not?

**Comment by northwestwitch** [comment]
Hi @Jakob37, I've written this PR addressing #357. You could perhaps test it if you want?

**Comment by Jakob37** [comment]
> Hi @Jakob37, I've written this PR addressing #357. You could perhaps test it if you want?

Great, I'll test!

File: src/chanjo2/meta/handle_d4.py [hunk 2]
```diff
@@ -282,3 +148,47 @@ def get_samples_sex_metrics(d4_file_path: str) -> Dict:
             x_cov=sex_chroms_coverage[0][1], y_cov=sex_chroms_coverage[1][1]
         ),
     }
+
+
+def get_gene_overview_stats(
+    sql_intervals: List[SQLTranscript],
+    samples: List[ReportQuerySample],
+    completeness_thresholds=List[int],
```
Comment (Jakob37) [inline src/chanjo2/meta/handle_d4.py:156]: Should it be:

```suggestion
    completeness_thresholds: List[int],
```

**Review by Jakob37 — COMMENTED** [review]
Hmm. I get some strange behaviour. I don't think it is due to the Chanjo2 code itself, but something in our setup. I wonder whether it is due to that we are accessing the d4 file over a network drive (`/fs1`).

First gene I opened took 15 seconds. Clear improvements.

Then I tried opening five, and things went downhill. Some opened in 3+ minutes, some timed out.

See here from htop (this can go on well beyond five minutes):

htop

~~I tried running the exact same command after first copying over the d4 file into the `/tmp` dir of the Chanjo2 container. And the command finished in <200ms.~~ Edit: Making sure I run on the exact same file it takes long time locally. Will have to test some more.

**Comment by Jakob37** [comment]
Not sure what is going on here 🤔 

```
$ time d4tools stat --region /tmp/chr1test <ID>-prune-dead-panels_coverage.d4 --stat mean
1       236664141       236720142       48.358297173264766
1       236686499       236764631       2717708.162443045
1       236686499       236764631       2717708.162443045
1       236686509       236763194       48.01204929256048
1       236686606       236764595       2722691.235174191
1       236686639       236764595       2723843.7670993893
1       236686639       236764595       2723843.7670993893
1       236686639       236764595       2723843.7670993893
1       236686659       236762714       47.923410689632504
1       236686674       236764595       2725067.2182723526
1       236686675       236764595       2725102.190220739
1       236686682       236764595       2725347.0191110596
1       236686689       236763120       2778190.6666666665
1       236686731       236762597       47.91945008304115
1       236686735       236720974       47.68512514968311
1       236706399       236764595       3648687.1129287235
1       236718904       236737202       50.287189856814955
1       236729087       236764595       5980002.298467951
1       236731250       236762848       47.459775935185775
1       236747700       236748140       48.3
1       236751199       236764603       15841301.848925693
1       236753812       236764603       19677202.179594107
1       236754254       236764595       20533474.795184217
1       236754898       236763035       40.139363401745115
1       236755382       236762909       39.087950046499266
```

**Comment by northwestwitch** [comment]
I'm going to do some testing on our stage server as well!

**Comment by northwestwitch** [comment]
All genes pages opened from out test server are extremely quick to open. Then I got this error for a gene named _**GAREM**_: 

I think it's because it found no gene named GAREM, since GAREM has changed name and now it's named GAREM1

```

[2024-09-30 08:56:09 +0200] [10] [ERROR] Exception in ASGI application
Traceback (most recent call last):
  File "/venv/lib/python3.11/site-packages/uvicorn/protocols/http/httptools_impl.py", line 436, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/fastapi/applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "/venv/lib/python3.11/site-packages/starlette/applications.py", line 123, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/venv/lib/python3.11/site-packages/starlette/middleware/errors.py", line 186, in __call__
    raise exc
  File "/venv/lib/python3.11/site-packages/starlette/middleware/errors.py", line 164, in __call__
    await self.app(scope, receive, _send)
  File "/venv/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 65, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "/venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/venv/lib/python3.11/site-packages/starlette/routing.py", line 756, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/venv/lib/python3.11/site-packages/starlette/routing.py", line 776, in app
    await route.handle(scope, receive, send)
  File "/venv/lib/python3.11/site-packages/starlette/routing.py", line 297, in handle
    await self.app(scope, receive, send)
  File "/venv/lib/python3.11/site-packages/starlette/routing.py", line 77, in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  File "/venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/venv/lib/python3.11/site-packages/starlette/routing.py", line 72, in app
    response = await func(request)
               ^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/fastapi/routing.py", line 278, in app
    raw_response = await run_endpoint_function(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/fastapi/routing.py", line 191, in run_endpoint_function
    return await dependant.call(**values)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/worker/app/src/chanjo2/endpoints/overview.py", line 106, in gene_overview
    get_gene_overview_coverage_stats(form_data=validated_form, session=db)
  File "/home/worker/app/src/chanjo2/meta/handle_report_contents.py", line 229, in get_gene_overview_coverage_stats
    gene_stats["samples_coverage_stats_by_interval"] = get_gene_overview_stats(
                                                       ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/worker/app/src/chanjo2/meta/handle_d4.py", line 181, in get_gene_overview_stats
    transcripts_completeness = get_d4tools_intervals_completeness(
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/worker/app/src/chanjo2/meta/handle_completeness_stats.py", line 15, in get_d4tools_intervals_completeness
    d4tools_stats_perc_cov: str = subprocess.check_output(
                                  ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/subprocess.py", line 466, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['d4tools', 'stat', '-s', 'perc_cov=10,15,20,50,100', '--region', '/tmp/tmpo4bywr8d', '/home/proj/stage/housekeeper-bundles/safeguinea/2023-06-26/hg002.d4']' died with <Signals.SIGSEGV: 11>.
```

**Comment by Jakob37** [comment]
I get crashes for d4tools when running locally on the master branch:

```
$ /home/jakob/src/d4-format/target/debug/d4tools stat --region debug.tsv <ID>-prune-dead-panels_coverage.d4 --stat mean
thread '<unnamed>' panicked at /home/jakob/src/d4-format/d4/src/d4file/track.rs:215:21:
attempt to subtract with overflow
```

For these regions:

```
1   236664141   236720142
1   236686499   236764631
1   236686509   236763194
1   236686606   236764595
1   236686639   236764595
1   236686659   236762714
1   236686674   236764595
1   236686675   236764595
1   236686682   236764595
1   236686689   236763120
1   236686731   236762597
1   236686735   236720974
1   236706399   236764595
1   236718904   236737202
1   236729087   236764595
1   236731250   236762848
1   236747700   236748140
1   236751199   236764603
1   236753812   236764603
1   236754254   236764595
1   236754898   236763035
1   236755382   236762909
```

Removing a single of them and it works 🤔

**Comment by northwestwitch** [comment]
> Do you have ACTN2 easily accessible to try out? This one is the one crashing for me.
> 
> Some others seems to load in ~10 seconds for me.

No, that gene works for me:

image

**Comment by Jakob37** [comment]
Hmm. Running on a separate d4 file (my GIAB file) also crashes. I'll poke around a bit in the code and see if I can find what is going wrong.

**Comment by northwestwitch** [comment]
> Hmm. Running on a separate d4 file (my GIAB file) also crashes. I'll poke around a bit in the code and see if I can find what is going wrong.

Are you using the latest master branch of d4tools?

**Comment by Jakob37** [comment]
> Are you using the latest master branch of d4tools?

Yes, I think so

d4tools_git
d4tools_github

I think this is a bug. Note that in the final range the order is reversed.

```
part_left: 236763120 part_right: 236763194
part_left: 236763194 part_right: 236764595
part_left: 236764595 part_right: 236763035
thread '<unnamed>' panicked at /home/jakob/src/d4-format/d4/src/d4file/track.rs:216:21:
attempt to subtract with overflow
```

For:

```
                println!("part_left: {} part_right: {}", part_left, part_right);
                decoder.decode_block(
                    part_left as usize,
                    (part_right - part_left) as usize,
                    block_handler,
                );
```

**Comment by northwestwitch** [comment]
I think #364 is another type of problem then!

**Comment by Jakob37** [comment]
> I think #364 is another type of problem then!

Yes, I think so too.

**Comment by Jakob37** [comment]
OK, it is probably not a super tricky bug though. It is in the same code that we previously have been working with. Maybe even a side effect of the previous changes.

I'll see if I can make another tentative fix and PR to the d4tools repo ...

**Comment by Jakob37** [comment]
OK, not so obvious how to fix it. I get it reliably for this set of ranges:

out

It has to do with how the right end is handled for the ranges.

But, I cannot easily reproduce it in a more narrow case. If I remove any of the ranges here, it does no longer lead to a crash :(

I'll dig around some more today. We'll see if I can figure it out.

But starting to think anyway. Do we have any backup more mature and better supported format that could be used instead of alternatively of d4 in Chanjo2? Like bigwig? I realize that is a big decision, but maybe something worth thinking about.

**Comment by northwestwitch** [comment]
> But starting to think anyway. Do we have any backup more mature and better supported format that could be used instead of alternatively of d4 in Chanjo2? Like bigwig? I realize that is a big decision, but maybe something worth thinking about.

Hehe the alternative is chanjo, which is using sambamba to evaluate coverage completeness.

**Comment by northwestwitch** [comment]
Now that the problem in d4tools should be solved I just need to refresh the docker base image used in the Dockerfile. I'm working on it!

**Comment by Jakob37** [comment]
> Now that the problem in d4tools should be solved I just need to refresh the docker base image used in the Dockerfile. I'm working on it!

Great! Happy to review when ready.

**Comment by northwestwitch** [comment]
Updated! My gene works fine now --->

image

**Comment by Jakob37** [comment]
> Updated! My gene works fine now --->
> image

Nice 🎉 I'll test right away

**Review by Jakob37 — APPROVED** [review]
Works well in testing. One question on the code changes, whether the ordering might be lost. Otherwise, looks great 👌

File: src/chanjo2/meta/handle_d4.py [hunk 3]
```diff
@@ -282,3 +148,47 @@ def get_samples_sex_metrics(d4_file_path: str) -> Dict:
...
+    samples: List[ReportQuerySample],
+    completeness_thresholds: List[int],
+) -> Dict[str, list]:
+    """Returns stats to be included in the gene overview page."""
+    interval_ids_coords: List[Tuple[str, Tuple[str, int, int]]] = [
+        (interval.ensembl_id, (interval.chromosome, interval.start, interval.stop))
+        for interval in sql_intervals
+    ]
+    interval_ids_coords = tuple(
+        set(sort_interval_ids_coords(interval_ids_coords))
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply src/chanjo2/meta/handle_d4.py:164]: That's a very good point: I'll just use list comprehension here. Thanks!

[the remark this answers is not in the document]
Comment (northwestwitch) [reply src/chanjo2/meta/handle_d4.py:164]: Or perhaps fiest set and then sorting

**Pull request merged**